mlir/lib/Analysis/Liveness.cpp

```diff
@@ -67,11 +67,18 @@ struct BlockInfoBuilder {
     // Mark all nested operation results as defined, and nested operation
     // operands as used. All defined value will be removed from the used set
     // at the end.
-    block->walk([&](Operation *op) {
-      for (Value result : op->getResults())
-        defValues.insert(result);
-      for (Value operand : op->getOperands())
-        useValues.insert(operand);
+    block->walk([&](Block *nestedBlock) {
+      if (block != nestedBlock) {
+        for (BlockArgument arg : nestedBlock->getArguments()) {
+          defValues.insert(arg);
+        }
+      }
+      for (Operation &op : *nestedBlock) {
```

Comment:
This makes it that we're traversing each block twice, seems unfortunate, any idea to do better?

Reply:
Yes, of course, I suggest the following option, but it will require an additional set:
```cpp
    SmallPtrSet<Block *, 32> visited;
    block->walk([&](Operation *op) {
      for (Value result : op->getResults())
          defValues.insert(result);
        for (Value operand : op->getOperands())
          useValues.insert(operand);
      
      Block *parent = op->getBlock();
      if (visited.contains(parent))
        return;
      visited.insert(parent);
      for (BlockArgument arg : parent->getArguments()) {
        defValues.insert(arg);
      }
    });
```

Reply:
What about visiting the child blocks for each operation instead of the parent block?

Reply:
I fixed it, please take a look when you get a chance